src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,224 @@
+    pub buffer_offset: Offset,
+
+    /// The type of acceleration structure to build.
+    pub ty: AccelerationStructureType,
+    // /// currently only has `accelerationStructureCaptureReplay`
+    // create_flags: VkAccelerationStructureCreateFlagsKHR,
+    // /// used for `accelerationStructureCaptureReplay`
+    // device_address: VkDeviceAddress,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkBuildAccelerationStructureFlagBitsKHR.html
+    pub struct AccelerationStructureFlags: u32 {
+        /// indicates that the specified acceleration structure can be updated with update of VK_TRUE in vkCmdBuildAccelerationStructuresKHR or vkCmdBuildAccelerationStructureNV .
+        const ALLOW_UPDATE = 0x1;
+        ///  indicates that the specified acceleration structure can act as the source for a copy acceleration structure command with mode of VK_COPY_ACCELERATION_STRUCTURE_MODE_COMPACT_KHR to produce a compacted acceleration structure.
+        const ALLOW_COMPACTION = 0x2;
+        /// indicates that the given acceleration structure build should prioritize trace performance over build time.
+        const PREFER_FAST_TRACE = 0x4;
+        ///  indicates that the given acceleration structure build should prioritize build time over trace performance.
+        const PREFER_FAST_BUILD = 0x8;
+        /// indicates that this acceleration structure should minimize the size of the scratch memory and the final result build, potentially at the expense of build time or trace performance.
+        const LOW_MEMORY = 0x10;
+    }
+}
+
+#[derive(Debug)]
+pub enum AccelerationStructureCreateMode {
+    /// specifies that the destination acceleration structure will be built using the specified geometries.
+    Build,
+    /// specifies that the destination acceleration structure will be built using data in a source acceleration structure, updated by the specified geometries.
+    /// Note there's constraints on update: https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#acceleration-structure-update-constraints
+    Update,
+}
+
+/// A description of the geometry data needed to populate an acceleration structure.
+///
+/// TODO: there's something here that smells w/ what fields are needed to get the required build size vs what fields are needed to actually build. Also, the top/bottom levels having different requirements on which fields are valid.
+// TODO: I don't like that this is reused for the "get build sizes" and "do actual build" with complex rules on what fields are ignored when. Perhaps we could use DX12's model for `D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC` and `D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS` and move at least the src/dst fields onto the actual "build acceleration structure" command.?
```

Comment:
This is the struct I'm most interested in feedback for. Mentioned in the many TODO comments on and around it, this is a kind of central type which most examples I've seen just keep around a mutable version of this and update fields as needed by methods.

I think we can move the src/dst/scratch values onto `build_acceleration_structures*` or have another type that has src/dst/scratch and a `AccelerationStructureGeometryDesc`.

Reply:
That sounds appropriate, let's try that!
I assume the Vulkan backend would be able to build this stuff internally with no issues if we separate it and follow DX12 style here?

Reply:
That should be correct.

Tangentially related: I'd wager we won't be able to reuse vulkan or dx structs across calls that effectively (i.e. the user will have a gfx *Desc, rather than a vulkan or dx *Desc struct), so I assume that we're okay with recreating these descriptor structs as needed?

Reply:
That's correct, yes. We'll be recreating them all the time.